lib/taste_tester/tunnel.rb

```diff
@@ -44,13 +38,17 @@ def run
     end
 
     def cmd
-      @max_ping = @delta_secs / 10
       pid = '$$'
       @ts = TasteTester::Config.testing_end_time.strftime('%y%m%d%H%M.%S')
+
+      # Tie the faith of our SSH tunnel with the faith of timestamp file.
+      # taste-testing can be renewed, so we'll wait until the timestamp file
+      # is entirely gone before we kill the tunnel
       cmds = "ps -o pgid= -p $(ps -o ppid= -p #{pid}) | sed \"s| ||g\" " +
              " > #{TasteTester::Config.timestamp_file} &&" +
              " touch -t #{@ts} #{TasteTester::Config.timestamp_file} &&" +
-             " sleep #{@delta_secs}"
+             "while [ -f #{TasteTester::Config.timestamp_file} ]; do " +
```

Comment:
this is race-y. If we try to tt again within those 60 seconds (which is likely if someone is taking over testing, they'll untest/test in rapid succession), this will break.

I wonder if we want to put our own PID into the timestamp file and have taste-untester updated to kill both?

Reply:
I'm less concerned about the tunnel outliving the test, I'm worried about the test outliving the tunnel.

In current implementation, Chef will fail. In the case you're describing, we may possibly have stale tunnels, which is not a big concern IMHO. Am I missing something here?

Since this is still better than the existing implementation, can we land this and create a followup?
As for your fix suggestion, I don't fully understand the lifecycle of a test. For example, is there just a single code path that deletes the timestamp file?

Reply:
Yes there is a single codepath that deletes the timestamp file - see taste-untester in this repo.

I think trading one bug for another isn't a good idea - we have non-Facebook users who may have worked around this bug, but aren't expecting long-lived stale tunnels (which is a potential security issue) on hosts.

If we put both PIDs in the timestamp file and then update taste-untester to do the right thing, I think that solves both problems.

Reply:
I don't follow the discussion here.  It looks like taste-untester already kills the PGID, which we already put in the file; is killing the PGID somehow not enough?

Reply:
Interesting point. So we take the pid of the shell we ssh'd in, get ITS pgid.

While that file is still there we sleep. In theory, the shell that's running the `while... sleep` is in the same process group, so in THEORY, if we kill pgid we should kill this shell and the tunnel. If the shell dies, that should be enough. Sure, I'm OK with that, if it works.

BTW, this has been modified since I last looked at this, I'm not sure what this line is supposed to be testing:

```
             "\"$(ps -o pgid= -p $(ps -o ppid= -p #{pid}) | sed \"s| ||g\")\" == \"$SSH_PGID\" ]]; do " +
```

making sure we didn't get re-parented is the idea there? I think we can keep our same pgid even if we get re-parented?